llap-common/src/java/org/apache/hadoop/hive/llap/metrics/ReadWriteLockMetrics.java

```diff
@@ -0,0 +1,428 @@
+        }
+
+        // sort by label (ascending) if lock time is the same
+        return lms1.lockLabel.compareTo(lms2.lockLabel);
+      }
+
+      return 0;
+    }
+  }
+
+  /**
+   * Wraps a <code>ReadWriteLock</code> into a monitored lock if required by
+   * configuration. This helper is checking the <code>
+   * hive.llap.lockmetrics.collect</code> configuration option and wraps the
+   * passed in <code>ReadWriteLock</code> into a monitoring container if the
+   * option is set to <code>true</code>. Otherwise, the original (passed in)
+   * lock instance is returned unmodified.
+   *
+   * @param conf Configuration instance to check for LLAP conf options
+   * @param lock The <code>ReadWriteLock</code> to wrap for monitoring
+   * @param metrics The target container for locking metrics
+   * @see #createLockMetricsSource
+   */
+  public static ReadWriteLock wrap(Configuration conf, ReadWriteLock lock, 
+                                   MetricsSource metrics) {
+    Preconditions.checkNotNull(lock, "Caller has to provide valid input lock");
+    boolean needsWrap = false;
+
+    if (null != conf) {
+      needsWrap = 
+        HiveConf.getBoolVar(conf, HiveConf.ConfVars.LLAP_COLLECT_LOCK_METRICS);
+    }
+
+    if (false == needsWrap) {
+      return lock;
+    }
+
+    Preconditions.checkNotNull(metrics, 
+        "Caller has to procide group specific metrics source");
+    return new ReadWriteLockMetrics(lock, metrics);
+  }
+
+  /**
+   * Factory method for new metric collections.
+   * You can create and use a single <code>MetricsSource</code> collection for
+   * multiple R/W locks. This makes sense if several locks belong to a single
+   * group and you're then interested in the accumulated values for the whole
+   * group, rather than the single lock instance. The passed in label is
+   * supposed to identify the group uniquely.
+   *
+   * @param label The group identifier for lock statistics
+   */
+  public static MetricsSource createLockMetricsSource(String label) {
+    Preconditions.checkNotNull(label);
+    Preconditions.checkArgument(!label.contains("\""), 
+        "Label can't contain quote (\")");
+    return new LockMetricSource(label);
+  }
+
+  /**
+   * Returns a list with all created <code>MetricsSource</code> instances for
+   * the R/W lock metrics. The returned list contains the instances that were
+   * previously created via the <code>createLockMetricsSource</code> function.
+   *
+   * @return A list of all R/W lock based metrics sources
+   */
+  public static List<MetricsSource> getAllMetricsSources() {
+    ArrayList<MetricsSource> ret = null;
+
+    synchronized (LockMetricSource.allInstances) {
+      ret = new ArrayList<>(LockMetricSource.allInstances);
+    }
+
+    return ret;
+  }
+
+  /// Enumeration of metric info names and descriptions
+  @VisibleForTesting
+  public enum LockMetricInfo implements MetricsInfo {
+    ReadLockWaitTimeTotal("The total wait time for read locks in nanoseconds"),
```

Comment:
minor nit i like when variable states the units like `ReadLockWaitTimeTotalNs`, but no big deal.